packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1970 @@
+   *
+   * In addition, this method will throw if the subject does not have a
+   * permission for the specified restricted method.
+   *
+   * @param origin - The origin of the subject to execute the method on behalf
+   * of.
+   * @param methodName - The name of the method to execute.
+   * @param params - The parameters to pass to the method implementation.
+   * @returns The result of the executed method.
+   */
+  async executeRestrictedMethod(
+    origin: OriginString,
+    methodName: ExtractPermission<
+      ControllerPermissionSpecification,
+      ControllerCaveatSpecification
+    >['parentCapability'],
+    params?: RestrictedMethodParameters,
+  ): Promise<Json> {
+    // Throws if the method does not exist
+    const methodImplementation = this._getRestrictedMethod(methodName, origin);
+
+    const result = await this._executeRestrictedMethod(
+      methodImplementation,
+      { origin },
+      methodName,
+      params,
+    );
+
+    if (result === undefined) {
+      throw new Error(
+        `Internal request for method "${methodName}" as origin "${origin}" returned no result.`,
+      );
+    }
+
+    return result;
+  }
+
+  /**
+   * An internal method used in the controller's `json-rpc-engine` middleware
+   * and {@link PermissionController.executeRestrictedMethod}. Calls the
+   * specified restricted method implementation after decorating it with the
+   * caveats of its permission. Throws if the subject does not have the
+   * requisite permission.
+   *
+   * **ATTN:** Assumes that all parameters are valid.
+   *
+   * @see {@link PermissionController.executeRestrictedMethod} and
+   * {@link PermissionController.createPermissionMiddleware} for usage.
+   * @param methodImplementation - The implementation of the method to call.
+   * @param subject - Metadata about the subject that made the request.
+   * @param req - The request object associated with the request.
+   * @returns
+   */
+  private _executeRestrictedMethod(
+    methodImplementation: RestrictedMethod<RestrictedMethodParameters, Json>,
+    subject: PermissionSubjectMetadata,
+    method: ExtractPermission<
+      ControllerPermissionSpecification,
+      ControllerCaveatSpecification
+    >['parentCapability'],
+    params: RestrictedMethodParameters = [],
+  ): ReturnType<RestrictedMethod<RestrictedMethodParameters, Json>> {
+    const { origin } = subject;
+
+    const permission = this.getPermission(origin, method);
+    if (!permission) {
+      throw unauthorized({ data: { origin, method } });
+    }
+
+    return decorateWithCaveats<ControllerCaveatSpecification>(
+      methodImplementation,
+      permission,
+      this._caveatSpecifications,
+    )({ method, params, context: { origin } });
+  }
+
+  /**
+   * Like {@link PermissionController.getRestrictedMethod}, except the request
+   * for the restricted method is associated with a specific subject, and a
+   * JSON-RPC error is thrown if the method does not exist.
```

Comment:
Any reason not to include this validation as part of `getRestrictedMethod`?

Reply:
As it turns out, no, and `_getRestrictedMethod` is redundant: fb667ff